sdk/python/lib/pulumi/provider/experimental/provider.py

```diff
@@ -63,32 +63,94 @@ def construct(
         self.validate_resource_type(self.metadata.name, resource_type)
         component_name = resource_type.split(":")[-1]
         comp = self.analyzer.find_component(self.path, component_name)
-        # Use the component definitions to map the schema names to the Python
-        # names.
-        # TODO: Handle complex types, including multiple levels of nesting.
         component_def = self._component_defs[component_name]
-        mapped_args = self.map_input_names(inputs, component_def.inputs_mapping)
+        mapped_args = self.map_inputs(inputs, component_def)
         # ComponentResource's init signature is different from the derived class signature.
         comp_instance = comp(name, mapped_args, options)  # type: ignore
-        return ConstructResult(
-            comp_instance.urn,
-            self.get_state(comp_instance, component_def.outputs_mapping),
-        )
+        state = self.get_state(comp_instance, component_def)
+        return ConstructResult(comp_instance.urn, state)
 
-    def map_input_names(self, inputs: Inputs, mapping: dict[str, str]) -> Inputs:
-        r: dict[str, Input[Any]] = {}
-        for k, v in inputs.items():
-            r[mapping[k]] = v
-        return r
+    def get_type_definition(self, prop: PropertyDefinition) -> TypeDefinition:
+        """Gets the type definition for a property with a type reference."""
+        if not prop.ref:
+            raise ValueError(f"property {prop} is not a complex type")
+        name = prop.ref.split(":")[-1]
+        return self._type_defs[name]
+
+    def map_inputs(self, inputs: Inputs, component_def: ComponentDefinition) -> Inputs:
```

Comment:
I want to add Python tests here, but I think I first need to make the schema analysis lazy like in Node.js, so we can create a Provider without running the whole analzyer. I'll do that in a follow up PR>